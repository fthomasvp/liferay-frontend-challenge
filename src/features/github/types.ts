export type TGitHubRepo = {
  id: number;
  full_name: string;
  created_at: string;
  stargazers_count: number;
  language: string;
  forks_count: number;
  open_issues_count: number;
  license: {
    name: string;
  } | null;
  lastCommitAt: string | null;
  isFavorited: boolean;
};

export type TGitHubCommit = {
  commit: {
    committer: {
      date: string;
      email: string;
      name: string;
    };
  };
};

export type TGitHubParams = {
  username: string;
  repositoryName: string;
};
