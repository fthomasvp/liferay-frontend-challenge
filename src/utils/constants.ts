export const MINIMUM_REPO_NAME_LENGTH = 1;
