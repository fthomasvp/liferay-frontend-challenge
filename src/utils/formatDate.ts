import { formatDistanceToNow } from 'date-fns';

export const formatDistanceTimeFromNow = (date: string) => {
  return formatDistanceToNow(new Date(date), {
    addSuffix: true,
  });
};
