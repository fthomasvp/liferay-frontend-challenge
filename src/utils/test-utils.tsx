import React, { ProviderProps, ReactElement } from 'react';
import { render } from '@testing-library/react';

import { HomeContext, THomeContext } from 'context/HomeContext';
import {
  SearchBarContext,
  SearchBarContextType,
} from 'context/search-bar/search-bar.context';

/**
 * A custom render to setup providers. Extends regular
 * render options with `providerProps` to allow injecting
 * different scenarios to test with.
 *
 * @see https://testing-library.com/docs/react-testing-library/setup#custom-render
 */
// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
const customRender = (
  ui: ReactElement,
  {
    homeProviderProps,
    searchBarProviderProps,
    ...renderOptions
  }: {
    homeProviderProps: ProviderProps<THomeContext>;
    searchBarProviderProps: ProviderProps<SearchBarContextType>;
  }
) => {
  return render(
    <HomeContext.Provider {...homeProviderProps}>
      <SearchBarContext.Provider {...searchBarProviderProps}>
        {ui}
      </SearchBarContext.Provider>
    </HomeContext.Provider>,
    renderOptions
  );
};

export * from '@testing-library/react';
export { customRender as render };
