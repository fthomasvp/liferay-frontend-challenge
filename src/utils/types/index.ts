export type GitHubRepo = {
  id: number;
  full_name: string;
  created_at: string;
  stargazers_count: number;
  language: string;
  forks_count: number;
  open_issues_count: number;
  license: {
    name: string;
  } | null;
  lastCommitAt: string | null;
  isFavorited: boolean;
};

export type Filter = {
  repositoryName?: string;
  isFavorited?: boolean;
};
