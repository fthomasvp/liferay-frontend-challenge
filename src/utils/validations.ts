/**
 * RegEx to validate repository full name.
 * The user must provide an input with at least
 * one word, followed by a slash and another word.
 */
export function isValidRepositoryName(repoName: string) {
  return repoName.match(/\w{1,}\/\w{1,}/);
}
