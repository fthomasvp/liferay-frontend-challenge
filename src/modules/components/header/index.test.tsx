import React from 'react';
import { render, cleanup, fireEvent, waitFor } from '@testing-library/react';

import Header from './index';
import { GitHubRepo } from 'utils/types';

describe('<Header />', () => {
  afterEach(cleanup);

  const repositories: GitHubRepo[] = [
    {
      id: 22377139,
      full_name: 'liferay/clay',
      created_at: '2014-07-29T11:07:36Z',
      stargazers_count: 156,
      language: 'HTML',
      forks_count: 185,
      open_issues_count: 55,
      license: {
        name: 'Other',
      },
      lastCommitAt: '2021-03-05T12:23:22Z',
    },
  ];
  const addRepository = jest.fn();

  test('should render correctly', () => {
    const props = {
      repositories,
      addRepository,
    };

    const { container } = render(<Header {...props} />);
    expect(container).toMatchSnapshot();
  });

  test('should display a Popover, a text input and two buttons', async () => {
    const props = {
      repositories,
      addRepository,
    };

    const { getByTestId } = render(<Header {...props} />);

    fireEvent.click(getByTestId('addButtonToolbarManagement'));

    const popover = await waitFor(() => getByTestId('popover'));
    expect(popover).toBeInTheDocument();

    const repositoryTextInput = getByTestId('repositoryTextInput');
    expect(repositoryTextInput).toBeInTheDocument();

    const popoverButtonActions = getByTestId('popoverButtonActions');
    expect(popoverButtonActions.children.length).toEqual(2);
  });
});
