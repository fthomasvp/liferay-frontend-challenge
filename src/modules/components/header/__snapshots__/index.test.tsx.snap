// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Header /> should render correctly 1`] = `
<div>
  <nav
    class="management-bar navbar navbar-expand-md management-bar-light"
  >
    <div
      class="container-fluid container-fluid-max-xl"
    >
      <ul
        class="navbar-nav"
      >
        <li
          class="nav-item"
          style="margin-right: 28px;"
        >
          <img
            alt="github logo"
            src="github.svg"
          />
        </li>
        <li
          class="nav-item"
        >
          <span
            class="text-capitalize text-secondary font-weight-semi-bold"
          >
            Github Compare
          </span>
        </li>
        <li
          class="nav-item"
        >
          <div
            class="dropdown"
          >
            <button
              class="dropdown-toggle nav-link btn btn-unstyled"
              style=""
              type="button"
            >
              <span
                class="navbar-breakpoint-down-d-none"
              >
                <span
                  class="navbar-text-truncate"
                >
                  Filter and order
                </span>
                <svg
                  class="lexicon-icon lexicon-icon-caret-bottom inline-item inline-item-after"
                  role="presentation"
                >
                  <use
                    xlink:href="#caret-bottom"
                  />
                </svg>
              </span>
              <span
                class="navbar-breakpoint-d-none"
              >
                <svg
                  class="lexicon-icon lexicon-icon-filter"
                  role="presentation"
                >
                  <use
                    xlink:href="#filter"
                  />
                </svg>
              </span>
            </button>
          </div>
        </li>
      </ul>
      <div
        class="navbar-form navbar-form-autofit navbar-overlay navbar-overlay-sm-down"
      >
        <div
          class="container-fluid container-fluid-max-xl"
        >
          <form
            role="search"
          >
            <div
              class="input-group"
            >
              <div
                class="input-group-item"
              >
                <input
                  aria-label="Search"
                  class="form-control form-control input-group-inset input-group-inset-after"
                  placeholder="e.g. liferay/clay"
                  type="text"
                />
                <span
                  class="input-group-inset-item input-group-inset-item-after"
                >
                  <button
                    class="navbar-breakpoint-d-none btn btn-monospaced btn-unstyled"
                    type="button"
                  >
                    <svg
                      class="lexicon-icon lexicon-icon-times"
                      role="presentation"
                    >
                      <use
                        xlink:href="#times"
                      />
                    </svg>
                  </button>
                  <button
                    class="btn btn-monospaced btn-unstyled"
                    type="submit"
                  >
                    <svg
                      class="lexicon-icon lexicon-icon-search"
                      role="presentation"
                    >
                      <use
                        xlink:href="#search"
                      />
                    </svg>
                  </button>
                </span>
              </div>
            </div>
          </form>
        </div>
      </div>
      <ul
        class="navbar-nav"
      >
        <li
          class="nav-item navbar-breakpoint-d-none"
        >
          <button
            class="nav-link nav-link-monospaced btn btn-unstyled"
            type="button"
          >
            <svg
              class="lexicon-icon lexicon-icon-search"
              role="presentation"
            >
              <use
                xlink:href="#search"
              />
            </svg>
          </button>
        </li>
        <li
          class="nav-item"
        >
          <button
            class="nav-link nav-link-monospaced btn btn-unstyled"
            type="button"
          >
            <svg
              class="lexicon-icon lexicon-icon-star-o"
              role="presentation"
            >
              <use
                xlink:href="#star-o"
              />
            </svg>
          </button>
        </li>
        <li
          class="nav-item"
        >
          <button
            class="nav-link nav-link-monospaced btn btn-unstyled"
            type="button"
          >
            <svg
              class="lexicon-icon lexicon-icon-adjust"
              role="presentation"
            >
              <use
                xlink:href="#adjust"
              />
            </svg>
          </button>
        </li>
        <li
          class="nav-item"
        >
          <div
            class="dropdown"
          >
            <button
              class="dropdown-toggle nav-link-monospaced nav-link btn btn-unstyled"
              style=""
              type="button"
            >
              <svg
                class="lexicon-icon lexicon-icon-cards2"
                role="presentation"
              >
                <use
                  xlink:href="#cards2"
                />
              </svg>
            </button>
          </div>
        </li>
        <li
          class="nav-item"
          style="margin-left: 24px;"
        >
          <button
            aria-label="Add button"
            class="btn btn-monospaced btn-primary"
            data-testid="addButtonToolbarManagement"
            type="button"
          >
            <svg
              class="lexicon-icon lexicon-icon-plus"
              role="presentation"
            >
              <use
                xlink:href="#plus"
              />
            </svg>
          </button>
        </li>
      </ul>
    </div>
  </nav>
</div>
`;
